### ISSUE TITLE
"x.y" is incompatible with "package.x.y" if letter case of argument does not match filesystem

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#6673]
**Describe the bug**
I have been getting the following error on and off over the last weeks (line numbers may be off since I minimize example files):

```
C:\Build\project\project\classes.py
  C:\Build\project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
    "classes.Class" is incompatible with "project.classes.Class" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

I was able to reproduce it using a minimal set of files this time; then moved the files to another directory to package it into an example, and it failed reproducing; then moved the files back, and it still failed reproducing. Maybe some cache is involved.

I am still packaging the files in case they are helpful, even though I cannot guarantee they reproduce the issue in a clean installation.

**Code or Screenshots**

`C:\Build\project\project\classes.py`
```python
from __future__ import annotations

from project.util.function import function

class Class:
    ...

function(Class())
```

`C:\Build\project\project\util\function.py`
```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from project.classes import Class

def function(instance: Class) -> None:
    print(instance)
```

In addition, I have two empty `__init__.py` files in `C:\Build\project\project` and in `C:\Build\project\project\util`, and I had installed the project as an editable install (under a different name, though).

Given the seemingly random nature of this problem, I wonder if the problem I see is expected or not.

**VS Code extension or command-line**
pyright-python 1.1.339 on command line, ran from `C:\Build\project`

**Comment by bersbersbers** [comment]
I just had the crazy situation that I had, in the same VS Code instance, two terminals side by side, with the same `venv` active, with the same environment variables (compared the output of `set`), and `pyright` would behave differently in the two terminals. The difference between the two terminals was the capitalization of the of the working directory. I replaced the project name in the output above, so maybe it is missing that important aspect.

This is how I can reproduce the issue:
- Put the above-mentioned files in `C:\Build\Project\project`. Be sure to get the capitalization right.
- Now change the working directory to `C:\Build\project`. I have no direct way of doing this, but runnnig `code C:\Build\project` and opening a terminal there achieves it.
- `pyright` reproduces the issue
- `cd .. && cd Project`
- `pyright` does not reproduce the issue

```
Microsoft Windows [Version 10.0.19045.3693]
(c) Microsoft Corporation. All rights reserved.

C:\Build\project>set PYRIGHT_PYTHON_FORCE_VERSION=latest

C:\Build\project>pyright
C:\Build\project\project\classes.py
  C:\Build\project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
    "classes.Class" is incompatible with "project.classes.Class" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 

C:\Build\project>cd ..

C:\Build>cd Project

C:\Build\Project>pyright
0 errors, 0 warnings, 0 informations 

C:\Build\Project>
```

**Comment by erictraut** [comment]
I'm not able to repro this on my Mac using the repro steps above. It might be specific to Windows.

Perhaps someone on the pylance team could try to repro. @debonte, @rchiodo

I'm not sure what it means for "the working directory" to not match the capitalization used in the file system. 

I don't know if we should spend much time on this. My inclination is to just say "don't do that".

**Comment by bersbersbers** [comment]
> I'm not able to repro this on my Mac using the repro steps above. It might be specific to Windows.

I am not surprised at all - Macs uses a case-sensitive file system, is that correct? I think this issue is specific to case-*in*sensitive file systems (read, mostly Windows).

> I'm not sure what it means for "the working directory" to not match the capitalization used in the file system.

Well, it means the current working directory is (reported by the shell as) `C:\Build\project` while the filesystem entry is `C:\Build\Project`. On Windows with a case-insensitive file system, that is the same folder, but `pyright` seems to treat them differently.

> I don't know if we should spend much time on this. My inclination is to just say "don't do that".

Let me kindly express my disagreement, based on three reasons (#3 being the strongest by far):
1. I did not intentionally cause this by mistyping commands - I had a couple of working shortcuts, which continued to work after I renamed `project` to `Project`. (And I think that's how a case-insensitive file system should behave.) The "incorrect" working directory was passed by VS Code. The "don't do that part" on the user side would be "renaming `project` to `Project`", and I think that should be allowed.

2. While I found it difficult initially to reproduce capitalization mismatches in `cmd.exe`, it is incredible easy in PowerShell:
   ```
   PS C:\Build> cd project
   PS C:\Build\project> pyright
   0 errors, 0 warnings, 0 informations
   
   PS C:\Build\project> cd ..
   PS C:\Build> cd Project
   PS C:\Build\Project> pyright   
   C:\Build\Project\project\classes.py
   C:\Build\Project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
       "classes.Class" is incompatible with "project.classes.Class" (reportGeneralTypeIssues)
   1 error, 0 warnings, 0 informations
   ```
   So capitalization mismatches in working directories should be considered something very normal in PowerShell.

3. This issue is *not* restricted to working directories:
   ```
   PS C:\Build> pyright project
   0 errors, 0 warnings, 0 informations 
   
   PS C:\Build> pyright Project
   C:\Build\Project\project\classes.py
     C:\Build\Project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
       "Project.project.classes.Class" is incompatible with "project.project.classes.Class" (reportGeneralTypeIssues)
   1 error, 0 warnings, 0 informations 
   ```
   nor PowerShell:
   ```
   Microsoft Windows [Version 10.0.19045.3693]
   (c) Microsoft Corporation. All rights reserved.
   
   (.venv) C:\Build>pyright project
   0 errors, 0 warnings, 0 informations 
   
   (.venv) C:\Build>pyright Project 
   C:\Build\Project\project\classes.py
     C:\Build\Project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
       "Project.project.classes.Class" is incompatible with "project.project.classes.Class" (reportGeneralTypeIssues)
   1 error, 0 warnings, 0 informations
   ```

In case you are wondering, that kind of issues are quite common, ironically, in Microsoft products. Two examples that I found:
- https://github.com/microsoft/vscode/issues/187974
- https://github.com/microsoft/vscode-pylint/issues/332

I still think this should be fixed.

**Comment by erictraut** [comment]
> Macs uses a case-sensitive file system, is that correct

No, Mac file systems are case-insensitive. Linux is typically case-sensitive.

**Comment by bersbersbers** [comment]
> Mac file systems are case-insensitive

Thanks, learned something today ;)

By the way,
- the `__init__.py` files are not required for reproduction
- I can reproduce this independent of `pyright-python` (using their `node.exe`, but not their packages):
```
C:\Build>"%USERPROFILE%\.cache\pyright-python\nodeenv\Scripts\node.exe" C:\Build\pyright\packages\pyright\index.js project
0 errors, 0 warnings, 0 informations 

C:\Build>"%USERPROFILE%\.cache\pyright-python\nodeenv\Scripts\node.exe" C:\Build\pyright\packages\pyright\index.js Project 
C:\Build\Project\project\classes.py
  C:\Build\Project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
    "Project.project.classes.Class" is incompatible with "project.project.classes.Class" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
@rchiodo, do you think that your recent URI changes are likely to address this? I think your main dev machine is a Windows system, right? Mind if I assign this to you to see if you can repro?

**Comment by bersbersbers** [comment]
I tested fd3c28f50d08b17f49c14a98943774d144b3be99 to see if

> recent URI changes are likely to address this

(by which you are referring to https://github.com/microsoft/pyright/pull/6519, I guess) and it looks like it does not:

```
(nodeenv) C:\Build>node pyright\packages\pyright\index.js project
0 errors, 0 warnings, 0 informations 

(nodeenv) C:\Build>node pyright\packages\pyright\index.js Project 
C:\Build\Project\project\classes.py
  C:\Build\Project\project\classes.py:10:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
    "Project.project.classes.Class" is incompatible with "project.project.classes.Class" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations

```

**Comment by rchiodo** [comment]
I can reproduce the problem using Powershell:

```
PS C:\Users\rchiodo\source\testing\Case> cd project
PS C:\Users\rchiodo\source\testing\Case\project> pyright
C:\Users\rchiodo\source\testing\Case\project\project\classes.py
  C:\Users\rchiodo\source\testing\Case\project\project\classes.py:8:10 - error: Argument of type "Class" cannot be assigned to parameter "instance" of type "Class" in function "function"
    "classes.Class" is incompatible with "project.classes.Class" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
PS C:\Users\rchiodo\source\testing\Case\project> cd ..
PS C:\Users\rchiodo\source\testing\Case> cd Project
PS C:\Users\rchiodo\source\testing\Case\Project> pyright
0 errors, 0 warnings, 0 informations
PS C:\Users\rchiodo\source\testing\Case\Project>
```

Although it seems to be the opposite case for me. Capital `Project` passes, while lower case `project` does not.

**Comment by rchiodo** [comment]
I cannot reproduce in VS code no matter where I start from though. Only on the command line when I force pyright to use the non-matching case.

**Comment by rchiodo** [comment]
Yeah, it consistently reproduces if I pass Pyright the wrong case. (regardless of what the true case is)

**Comment by erictraut** [comment]
That's good news that you're able to repro it. Let me know if you have any thoughts about a fix.

**Comment by bersbersbers** [comment]
> I cannot reproduce in VS code no matter where I start from though.

Let me know when you think I should help reproducing it in VS Code. (I think reproducing in PowerShell is good enough.)

**Comment by rchiodo** [comment]
@bersbersbers how did you verify the latest? Did you sync to main?

**Comment by bersbersbers** [comment]
@rchiodo oh, you are right! I did essentially the same thing except for `npm run build:cli:dev`, which I must have forgotten (I rarely use `npm` at all, and somehow `npm install` seemed to do what I expected). Now this is working fine for me, and I am looking forward to this Wednesday's build :)

**Comment by erictraut** [comment]
Ah, that's great news. Thanks @rchiodo for the repro and @bersbersbers for confirming the fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.340, which I just published. It will be included in a future release of pylance.
